cabal-install/Distribution/Client/CmdRepl.hs

```diff
@@ -100,62 +212,223 @@ replCommand = Client.installCommand {
 -- For more details on how this works, see the module
 -- "Distribution.Client.ProjectOrchestration"
 --
-replAction :: (ConfigFlags, ConfigExFlags, InstallFlags, HaddockFlags, [String])
+replAction :: (ConfigFlags, ConfigExFlags, InstallFlags, HaddockFlags, ReplFlags, EnvFlags)
            -> [String] -> GlobalFlags -> IO ()
-replAction (configFlags, configExFlags, installFlags, haddockFlags, replArgs)
+replAction (configFlags, configExFlags, installFlags, haddockFlags, replFlags, envFlags)
            targetStrings globalFlags = do
-
-    baseCtx <- establishProjectBaseContext verbosity cliConfig
-
-    targetSelectors <- either (reportTargetSelectorProblems verbosity) return
-                   =<< readTargetSelectors (localPackages baseCtx) targetStrings
-
-    buildCtx' <-
-      runProjectPreBuildPhase verbosity baseCtx $ \elaboratedPlan -> do
-
-            when (buildSettingOnlyDeps (buildSettings baseCtx)) $
-              die' verbosity $ "The repl command does not support '--only-dependencies'. "
-                 ++ "You may wish to use 'build --only-dependencies' and then "
-                 ++ "use 'repl'."
-
-            -- Interpret the targets on the command line as repl targets
-            -- (as opposed to say build or haddock targets).
-            targets <- either (reportTargetProblems verbosity) return
-                     $ resolveTargets
-                         selectPackageTargets
-                         selectComponentTarget
-                         TargetProblemCommon
-                         elaboratedPlan
-                         Nothing
-                         targetSelectors
-
-            -- Reject multiple targets, or at least targets in different
-            -- components. It is ok to have two module/file targets in the
-            -- same component, but not two that live in different components.
-            when (Set.size (distinctTargetComponents targets) > 1) $
-              reportTargetProblems verbosity
-                [TargetProblemMultipleTargets targets]
-
-            let elaboratedPlan' = pruneInstallPlanToTargets
-                                    TargetActionRepl
-                                    targets
-                                    elaboratedPlan
-            return (elaboratedPlan', targets)
+    let
+      onlySpecified = fromFlagOrDefault False (envOnlySpecified envFlags)
+      with    = withProject    cliConfig verbosity targetStrings
+      without = withoutProject cliConfig verbosity targetStrings
+    
+    (baseCtx, targetSelectors, finalizer) <- 
+      if onlySpecified
+        then 
+          without
+        else
+          catch with
+            $ \case
+              (BadPackageLocations prov locs) 
+                | prov == Set.singleton Implicit
+                , let 
+                  isGlobErr (BadLocGlobEmptyMatch _) = True
+                  isGlobErr _ = False
+                , any isGlobErr locs ->
```

Comment:
Code golf: can also be written as `notNull [ () | BadLocGlobEmptyMatch _ <- locs ]`.